File: server/app/utils/pdfGenerator.py
"""
PDF generation utilities for medical reports
"""
from reportlab.pdfgen import canvas
from reportlab.lib.pagesizes import letter, A4
from reportlab.lib.colors import HexColor, black, blue, green, red
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Table, TableStyle
from reportlab.lib.units import inch
from reportlab.lib import colors
from io import BytesIO
import os
from datetime import datetime
from typing import Dict, List, Optional, Any


def generateMedicalPDF(result_data: Dict[str, Any], patient_data: Dict[str, Any]) -> bytes:
    """
    Generate a professional medical report PDF
    
    Args:
        result_data: Analysis result data
        patient_data: Patient information
        
    Returns:
        bytes: PDF file content
    """
    buffer = BytesIO()
    
    # Create the PDF document
    doc = SimpleDocTemplate(
        buffer,
        pagesize=letter,
        rightMargin=72,
        leftMargin=72,
        topMargin=72,
        bottomMargin=18
    )
    
    # Container for the 'Flowable' objects
    story = []
    
    # Define styles
    styles = getSampleStyleSheet()
    
    # Custom styles
    title_style = ParagraphStyle(
        'CustomTitle',
        parent=styles['Title'],
        fontSize=18,
        spaceAfter=30,
        textColor=HexColor('#1a365d'),
        alignment=1  # Center alignment
    )
    
    heading_style = ParagraphStyle(
        'CustomHeading',
        parent=styles['Heading2'],
        fontSize=14,
        spaceAfter=12,
        textColor=HexColor('#2d3748'),
        borderWidth=1,
        borderColor=HexColor('#e2e8f0'),
        borderPadding=8,
        backColor=HexColor('#f7fafc')
    )
    
    normal_style = styles['Normal']
    normal_style.fontSize = 11
    normal_style.spaceAfter = 6
    
    # Title
    story.append(Paragraph("MEDICAL ANALYSIS REPORT", title_style))
    story.append(Spacer(1, 20))
    
    # Patient Information Section
    story.append(Paragraph("PATIENT INFORMATION", heading_style))
    
    patient_info = [
        ['Patient Name:', patient_data.get('full_name', 'N/A')],
        ['Date of Birth:', patient_data.get('date_of_birth', 'N/A')],
        ['Gender:', patient_data.get('gender', 'N/A')],
        ['Phone:', patient_data.get('phone', 'N/A')],
        ['Email:', patient_data.get('email', 'N/A')],
    ]
    
    patient_table = Table(patient_info, colWidths=[2*inch, 4*inch])
    patient_table.setStyle(TableStyle([
        ('BACKGROUND', (0, 0), (0, -1), HexColor('#edf2f7')),
        ('TEXTCOLOR', (0, 0), (-1, -1), black),
        ('ALIGN', (0, 0), (-1, -1), 'LEFT'),
        ('FONTNAME', (0, 0), (0, -1), 'Helvetica-Bold'),
        ('FONTNAME', (1, 0), (1, -1), 'Helvetica'),
        ('FONTSIZE', (0, 0), (-1, -1), 10),
        ('BOTTOMPADDING', (0, 0), (-1, -1), 8),
        ('TOPPADDING', (0, 0), (-1, -1), 8),
        ('GRID', (0, 0), (-1, -1), 1, HexColor('#e2e8f0'))
    ]))
    
    story.append(patient_table)
    story.append(Spacer(1, 20))
    
    # Analysis Results Section
    story.append(Paragraph("ANALYSIS RESULTS", heading_style))
    
    analysis_info = [
        ['Analysis Type:', get_analysis_type_display(result_data.get('model_type', 'unknown'))],
        ['Result:', result_data.get('prediction', 'N/A')],
        ['Confidence Score:', f"{(result_data.get('confidence', 0) * 100):.1f}%"],
        ['Analysis Date:', format_datetime(result_data.get('created_at'))],
        ['Model Used:', get_model_name(result_data.get('model_type', 'unknown'))],
        ['Status:', result_data.get('status', 'Completed').title()],
    ]
    
    if result_data.get('entropy'):
        analysis_info.append(['Entropy Score:', f"{result_data.get('entropy'):.3f}"])
    
    analysis_table = Table(analysis_info, colWidths=[2*inch, 4*inch])
    analysis_table.setStyle(TableStyle([
        ('BACKGROUND', (0, 0), (0, -1), HexColor('#edf2f7')),
        ('TEXTCOLOR', (0, 0), (-1, -1), black),
        ('ALIGN', (0, 0), (-1, -1), 'LEFT'),
        ('FONTNAME', (0, 0), (0, -1), 'Helvetica-Bold'),
        ('FONTNAME', (1, 0), (1, -1), 'Helvetica'),
        ('FONTSIZE', (0, 0), (-1, -1), 10),
        ('BOTTOMPADDING', (0, 0), (-1, -1), 8),
        ('TOPPADDING', (0, 0), (-1, -1), 8),
        ('GRID', (0, 0), (-1, -1), 1, HexColor('#e2e8f0'))
    ]))
    
    story.append(analysis_table)
    story.append(Spacer(1, 20))
    
    # Clinical Findings Section
    story.append(Paragraph("CLINICAL FINDINGS", heading_style))
    
    findings = generate_findings(result_data)
    for i, finding in enumerate(findings, 1):
        finding_text = f"{i}. {finding}"
        story.append(Paragraph(finding_text, normal_style))
    
    story.append(Spacer(1, 20))
    
    # Medical Recommendations Section
    story.append(Paragraph("MEDICAL RECOMMENDATIONS", heading_style))
    
    recommendations = generate_recommendations(result_data)
    for i, rec in enumerate(recommendations, 1):
        rec_text = f"{i}. {rec}"
        story.append(Paragraph(rec_text, normal_style))
    
    story.append(Spacer(1, 20))
    
    # Additional Notes
    if result_data.get('notes'):
        story.append(Paragraph("ADDITIONAL NOTES", heading_style))
        story.append(Paragraph(result_data['notes'], normal_style))
        story.append(Spacer(1, 20))
    
    # Disclaimer
    story.append(Paragraph("IMPORTANT MEDICAL NOTICE", heading_style))
    disclaimer_text = """
    This AI analysis is designed to assist healthcare professionals and should not replace 
    professional medical diagnosis. Please consult with a qualified healthcare provider 
    for proper medical evaluation and treatment decisions. This report is generated by 
    Second Opinion AI - Medical Analysis Platform.
    """
    
    disclaimer_style = ParagraphStyle(
        'Disclaimer',
        parent=normal_style,
        textColor=HexColor('#744210'),
        backColor=HexColor('#fef5e7'),
        borderWidth=1,
        borderColor=HexColor('#d69e2e'),
        borderPadding=10
    )
    
    story.append(Paragraph(disclaimer_text.strip(), disclaimer_style))
    
    # Footer
    story.append(Spacer(1, 30))
    footer_style = ParagraphStyle(
        'Footer',
        parent=normal_style,
        fontSize=8,
        textColor=HexColor('#718096'),
        alignment=1  # Center alignment
    )
    
    footer_text = f"Generated by Second Opinion AI - Medical Analysis Platform | {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}"
    story.append(Paragraph(footer_text, footer_style))
    
    # Build PDF
    doc.build(story)
    
    # Get the value of the BytesIO buffer
    pdf_content = buffer.getvalue()
    buffer.close()
    
    return pdf_content


def get_analysis_type_display(model_type: str) -> str:
    """Get display name for analysis type"""
    type_mapping = {
        'tumor': 'Brain Tumor Detection',
        'chest_xray': 'Pneumonia Detection',
        'unknown': 'Medical Analysis'
    }
    return type_mapping.get(model_type, 'Medical Analysis')


def get_model_name(model_type: str) -> str:
    """Get model name for display"""
    model_mapping = {
        'tumor': 'BrainTumor ViT v3.2',
        'chest_xray': 'ChestXray ViT v2.1',
        'unknown': 'Medical AI Model'
    }
    return model_mapping.get(model_type, 'Medical AI Model')


def format_datetime(dt_string: Optional[str]) -> str:
    """Format datetime string for display"""
    if not dt_string:
        return datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    
    try:
        # Try to parse ISO format
        dt = datetime.fromisoformat(dt_string.replace('Z', '+00:00'))
        return dt.strftime('%Y-%m-%d %H:%M:%S')
    except:
        # Return as-is if parsing fails
        return str(dt_string)


def generate_findings(result_data: Dict[str, Any]) -> List[str]:
    """Generate clinical findings based on analysis results"""
    prediction = result_data.get('prediction', '').lower()
    confidence = result_data.get('confidence', 0) * 100
    
    findings = []
    
    if 'normal' in prediction:
        findings.extend([
            "No significant abnormalities detected",
            "Normal tissue structure observed",
            "Clear anatomical features",
            "No mass effect or suspicious lesions identified"
        ])
    elif 'glioma' in prediction:
        findings.extend([
            f"{result_data.get('prediction', 'Glioma')} detected",
            "Irregular enhancement pattern observed",
            "Possible mass effect present",
            "Requires further neurological evaluation"
        ])
    elif 'meningioma' in prediction:
        findings.extend([
            f"{result_data.get('prediction', 'Meningioma')} detected",
            "Extra-axial location likely",
            "Well-circumscribed appearance",
            "Characteristic imaging features present"
        ])
    elif 'pituitary' in prediction:
        findings.extend([
            f"{result_data.get('prediction', 'Pituitary tumor')} detected",
            "Sellar/suprasellar region involvement",
            "Endocrine evaluation may be required",
            "Specialized imaging recommended"
        ])
    elif 'pneumonia' in prediction:
        findings.extend([
            "Pneumonia-like changes detected",
            "Consolidation or infiltrates present",
            "Respiratory system involvement",
            "Clinical correlation recommended"
        ])
    else:
        findings.extend([
            f"Analysis result: {result_data.get('prediction', 'Unknown')}",
            f"AI confidence level: {confidence:.1f}%",
            "Medical review recommended for interpretation",
            "Consider additional diagnostic studies if clinically indicated"
        ])
    
    # Add confidence-based findings
    if confidence >= 90:
        findings.append("High confidence in AI analysis results")
    elif confidence >= 75:
        findings.append("Moderate confidence in AI analysis results")
    else:
        findings.append("Low confidence - manual review strongly recommended")
    
    return findings


def generate_recommendations(result_data: Dict[str, Any]) -> List[str]:
    """Generate medical recommendations based on analysis results"""
    prediction = result_data.get('prediction', '').lower()
    
    recommendations = []
    
    if 'normal' in prediction:
        recommendations.extend([
            "Continue routine monitoring as clinically indicated",
            "No immediate intervention required based on imaging",
            "Follow standard surveillance protocols",
            "Consult healthcare provider for ongoing care plan"
        ])
    elif any(tumor in prediction for tumor in ['glioma', 'meningioma', 'pituitary', 'tumor']):
        recommendations.extend([
            "Immediate neurology or neurosurgery consultation recommended",
            "Additional imaging studies may be required (MRI with contrast)",
            "Multidisciplinary team review suggested",
            "Consider referral to specialized cancer center",
            "Discuss treatment options with healthcare team"
        ])
    elif 'pneumonia' in prediction:
        recommendations.extend([
            "Pulmonology consultation may be beneficial",
            "Consider antimicrobial therapy as appropriate",
            "Follow-up imaging to monitor treatment response",
            "Clinical correlation with symptoms and laboratory results"
        ])
    else:
        recommendations.extend([
            "Consult with healthcare provider for result interpretation",
            "Consider additional diagnostic studies as needed",
            "Clinical correlation with patient symptoms",
            "Follow-up as directed by healthcare team"
        ])
    
    # General recommendations
    recommendations.extend([
        "Maintain regular follow-up appointments",
        "Report any new or worsening symptoms immediately",
        "Keep this report for medical records"
    ])
    
    return recommendations

File: client/src/utils/pdfGenerator.js
import jsPDF from "jspdf";
import { formatNepaliTime, calculateAgeInNepali } from "./formatters";

/**
 * Generate a medical analysis PDF report
 * @param {Object} result - The analysis result data
 * @returns {Promise<jsPDF>} The generated PDF document
 */
export const generateMedicalPDF = async (result) => {
  try {
    const pdf = new jsPDF("p", "mm", "a4");
    const pageWidth = pdf.internal.pageSize.getWidth();
    const pageHeight = pdf.internal.pageSize.getHeight();
    let yPosition = 20;
    const margin = 20;
    const contentWidth = pageWidth - margin * 2;

    // Helper function to add a new page if needed
    const checkPageBreak = (requiredSpace = 30) => {
      if (yPosition > pageHeight - requiredSpace) {
        pdf.addPage();
        addPageHeader();
        yPosition = 45;
        return true;
      }
      return false;
    };

    // Helper function to add page header (for subsequent pages)
    const addPageHeader = () => {
      pdf.setFillColor(245, 245, 245);
      pdf.rect(0, 0, pageWidth, 25, "F");
      pdf.setTextColor(60, 60, 60);
      pdf.setFontSize(10);
      pdf.setFont(undefined, "normal");
      pdf.text(
        "SECOND OPINION AI - MEDICAL ANALYSIS REPORT",
        pageWidth / 2,
        15,
        { align: "center" }
      );
      pdf.setDrawColor(200, 200, 200);
      pdf.line(0, 25, pageWidth, 25);
      pdf.setTextColor(0, 0, 0);
    };

    // Helper function to add section divider
    const addSectionDivider = () => {
      pdf.setDrawColor(100, 100, 100);
      pdf.setLineWidth(0.3);
      pdf.line(margin, yPosition, pageWidth - margin, yPosition);
      yPosition += 5;
    };

    // Header - Navy blue background like the image
    pdf.setFillColor(52, 73, 121); // Navy blue color from the image
    pdf.rect(0, 0, pageWidth, 50, "F");

    pdf.setTextColor(255, 255, 255);
    pdf.setFontSize(20);
    pdf.setFont(undefined, "bold");
    pdf.text("MEDICAL IMAGING ANALYSIS REPORT", pageWidth / 2, 18, {
      align: "center",
    });

    pdf.setFontSize(11);
    pdf.setFont(undefined, "normal");
    pdf.text(
      "AI-Assisted Diagnostic Imaging • Second Opinion Medical AI Platform",
      pageWidth / 2,
      30,
      { align: "center" }
    );
    pdf.text(
      "This report should be reviewed by a qualified radiologist",
      pageWidth / 2,
      40,
      { align: "center" }
    );

    pdf.setTextColor(0, 0, 0);
    yPosition = 65;

    // I. PATIENT DEMOGRAPHICS
    pdf.setFontSize(12);
    pdf.setFont(undefined, "bold");
    pdf.setTextColor(52, 73, 121);
    pdf.text("I. PATIENT DEMOGRAPHICS", margin, yPosition);
    pdf.setTextColor(0, 0, 0);
    yPosition += 8;

    addSectionDivider();

    // Patient info table with proper formatting
    pdf.setFontSize(10);

    // Row 1
    pdf.setFont(undefined, "bold");
    pdf.text("Patient Name:", margin + 5, yPosition);
    pdf.setFont(undefined, "normal");
    pdf.text(
      result.patientName || result.patient?.full_name || "yojal",
      margin + 50,
      yPosition
    );

    pdf.setFont(undefined, "bold");
    pdf.text("Patient ID:", margin + 110, yPosition);
    pdf.setFont(undefined, "normal");
    pdf.text(result.patientId || "PT-000013", margin + 145, yPosition);
    yPosition += 10;

    // Row 2
    pdf.setFont(undefined, "bold");
    pdf.text("Date of Birth:", margin + 5, yPosition);
    pdf.setFont(undefined, "normal");
    const dob = result.patient?.date_of_birth
      ? formatNepaliTime(result.patient.date_of_birth, {
          hour: undefined,
          minute: undefined,
          timeZoneName: undefined,
        })
      : result.dob || "5/3/2003";
    pdf.text(dob, margin + 50, yPosition);

    pdf.setFont(undefined, "bold");
    pdf.text("Age:", margin + 110, yPosition);
    pdf.setFont(undefined, "normal");
    const age =
      result.age ||
      (result.patient?.date_of_birth
        ? `${calculateAgeInNepali(result.patient.date_of_birth)} years`
        : "22 years");
    pdf.text(age, margin + 145, yPosition);
    yPosition += 10;

    // Row 3
    pdf.setFont(undefined, "bold");
    pdf.text("Gender:", margin + 5, yPosition);
    pdf.setFont(undefined, "normal");
    pdf.text(
      result.patient?.gender || result.gender || "Male",
      margin + 50,
      yPosition
    );

    pdf.setFont(undefined, "bold");
    pdf.text("Contact:", margin + 110, yPosition);
    pdf.setFont(undefined, "normal");
    const contact =
      result.patient?.phone && result.patient.phone !== "N/A"
        ? result.patient.phone
        : result.contact && result.contact !== "N/A"
        ? result.contact
        : "986-786-9502";
    pdf.text(contact, margin + 145, yPosition);

    yPosition += 20;
    checkPageBreak(50);

    // II. IMAGING FINDINGS (renamed from III to II after removing Clinical Information)
    pdf.setFontSize(12);
    pdf.setFont(undefined, "bold");
    pdf.setTextColor(52, 73, 121);
    pdf.text("II. IMAGING FINDINGS", margin, yPosition);
    pdf.setTextColor(0, 0, 0);
    yPosition += 8;

    addSectionDivider();

    // Primary Finding Box - Blue border like in the image
    pdf.setDrawColor(52, 73, 121);
    pdf.setLineWidth(1);
    pdf.rect(margin, yPosition, contentWidth, 25, "D");

    pdf.setFontSize(11);
    pdf.setFont(undefined, "bold");
    pdf.setTextColor(52, 73, 121);
    pdf.text("PRIMARY FINDING:", margin + 10, yPosition + 8);

    pdf.setTextColor(52, 73, 121);
    pdf.text("AI Confidence:", pageWidth - margin - 50, yPosition + 8);

    // Result text
    pdf.setFontSize(14);
    pdf.setFont(undefined, "bold");
    pdf.setTextColor(0, 0, 0);
    const resultText = result.result || result.prediction || "PITUITARY TUMOR";
    pdf.text(resultText, margin + 10, yPosition + 18);

    // Confidence with colored background
    const rawConfidence = result.confidence || 0.74; // Default to 74% if not provided
    // Check if confidence is already a percentage (>1) or a decimal (0-1)
    const confidence =
      rawConfidence > 1
        ? Math.round(rawConfidence)
        : Math.round(rawConfidence * 100);
    pdf.setFontSize(12);
    pdf.setFont(undefined, "bold");

    // Green background for confidence like in the image
    pdf.setFillColor(76, 175, 80);
    pdf.rect(pageWidth - margin - 45, yPosition + 12, 35, 8, "F");
    pdf.setTextColor(255, 255, 255);
    pdf.text(`${confidence}%`, pageWidth - margin - 42, yPosition + 18);
    pdf.setTextColor(0, 0, 0);

    pdf.setFontSize(9);
    pdf.text("(HIGH)", pageWidth - margin - 25, yPosition + 18);

    yPosition += 35;

    // DETAILED OBSERVATIONS
    pdf.setFontSize(11);
    pdf.setFont(undefined, "bold");
    pdf.text("DETAILED OBSERVATIONS:", margin, yPosition);
    yPosition += 8;

    // Generate findings based on result or use provided ones
    const findings =
      result.findings && result.findings.length > 0
        ? result.findings
        : generateDefaultFindings(resultText);

    pdf.setFontSize(10);
    pdf.setFont(undefined, "normal");
    findings.forEach((finding, index) => {
      checkPageBreak(8);
      pdf.text(`${index + 1}.`, margin + 5, yPosition);
      pdf.text(finding, margin + 15, yPosition);
      yPosition += 8;
    });

    yPosition += 10;
    checkPageBreak(50);

    // III. CLINICAL IMPRESSION & RECOMMENDATIONS
    pdf.setFontSize(12);
    pdf.setFont(undefined, "bold");
    pdf.setTextColor(52, 73, 121);
    pdf.text("III. CLINICAL IMPRESSION & RECOMMENDATIONS", margin, yPosition);
    pdf.setTextColor(0, 0, 0);
    yPosition += 8;

    addSectionDivider();

    // Generate recommendations
    const recommendations =
      result.recommendations && result.recommendations.length > 0
        ? result.recommendations
        : generateDefaultRecommendations(resultText);

    pdf.setFontSize(10);
    pdf.setFont(undefined, "normal");
    recommendations.forEach((rec, index) => {
      checkPageBreak(8);
      pdf.text(`${index + 1}.`, margin + 5, yPosition);
      pdf.text(rec, margin + 15, yPosition);
      yPosition += 8;
    });

    yPosition += 15;
    checkPageBreak(60);

    // IMPORTANT MEDICAL DISCLAIMER - Red border box
    pdf.setDrawColor(220, 53, 69);
    pdf.setLineWidth(1);
    pdf.rect(margin, yPosition, contentWidth, 45, "D");

    pdf.setFillColor(254, 242, 242);
    pdf.rect(margin + 1, yPosition + 1, contentWidth - 2, 43, "F");

    pdf.setFontSize(11);
    pdf.setFont(undefined, "bold");
    pdf.setTextColor(220, 53, 69);
    pdf.text("⚠ IMPORTANT MEDICAL DISCLAIMER", margin + 10, yPosition + 10);

    pdf.setFontSize(9);
    pdf.setFont(undefined, "normal");
    pdf.setTextColor(0, 0, 0);
    const disclaimerText = `This AI-generated analysis is intended for use by qualified healthcare professionals only and should not be used as the sole basis for clinical decision-making. The results must be interpreted in conjunction with clinical findings, patient history, and other diagnostic information. In cases where there is uncertainty or clinical concern, additional imaging, consultation, or follow-up may be warranted. Emergency cases require immediate clinical consultation regardless of AI analysis results.`;

    const disclaimerLines = pdf.splitTextToSize(
      disclaimerText,
      contentWidth - 20
    );
    let disclaimerY = yPosition + 18;
    disclaimerLines.forEach((line) => {
      pdf.text(line, margin + 10, disclaimerY);
      disclaimerY += 5;
    });

    // Check if we need a new page for footer
    yPosition += 55;
    if (yPosition > pageHeight - 60) {
      pdf.addPage();
      yPosition = 40;
    }

    // Footer section - Signature area
    yPosition = pageHeight - 50;

    // Signature lines
    pdf.setDrawColor(0, 0, 0);
    pdf.setLineWidth(0.5);
    pdf.line(margin, yPosition, margin + 70, yPosition);
    pdf.line(pageWidth - margin - 70, yPosition, pageWidth - margin, yPosition);

    pdf.setFontSize(9);
    pdf.setFont(undefined, "normal");
    pdf.text("Reviewing Radiologist", margin, yPosition + 6);
    pdf.text("Date", pageWidth - margin - 70, yPosition + 6);

    pdf.setFontSize(8);
    pdf.setTextColor(100, 100, 100);
    pdf.text("(To be completed by licensed physician)", margin, yPosition + 12);

    // Page footer
    yPosition = pageHeight - 20;
    pdf.setDrawColor(200, 200, 200);
    pdf.line(margin, yPosition - 5, pageWidth - margin, yPosition - 5);

    pdf.setFontSize(8);
    pdf.text(
      "Second Opinion AI Medical Platform | Advanced Diagnostic Imaging Analysis",
      margin,
      yPosition
    );
    pdf.text(
      `Generated: ${formatNepaliTime(new Date())}`,
      pageWidth - margin,
      yPosition,
      { align: "right" }
    );
    pdf.text(
      "This report requires review and interpretation by a licensed radiologist",
      pageWidth / 2,
      yPosition + 6,
      { align: "center" }
    );
    pdf.text("Page 1", pageWidth - margin, yPosition + 12, { align: "right" });

    return pdf;
  } catch (error) {
    console.error("Error generating PDF:", error);
    throw error;
  }
};

// Helper function to generate default findings based on result
const generateDefaultFindings = (result) => {
  const resultLower = result.toLowerCase();

  if (resultLower.includes("pituitary")) {
    return [
      "Pituitary Tumor detected",
      "Sellar/suprasellar region",
      "Endocrine evaluation needed",
      "MRI with gadolinium recommended",
    ];
  } else if (resultLower.includes("glioma")) {
    return [
      "Glioma tumor detected",
      "Irregular enhancement pattern",
      "Possible mass effect",
      "Requires further evaluation",
    ];
  } else if (resultLower.includes("meningioma")) {
    return [
      "Meningioma tumor detected",
      "Extra-axial location",
      "Well-circumscribed borders",
      "Dural enhancement present",
    ];
  } else if (resultLower.includes("normal")) {
    return [
      "No tumor detected",
      "Normal brain tissue structure",
      "Clear ventricles",
      "No mass effect observed",
    ];
  } else {
    return [
      `${result} detected`,
      "Requires clinical correlation",
      "Further evaluation recommended",
      "Follow-up imaging may be needed",
    ];
  }
};

// Helper function to generate default recommendations
const generateDefaultRecommendations = (result) => {
  const resultLower = result.toLowerCase();

  if (resultLower.includes("normal")) {
    return [
      "Continue routine monitoring",
      "No immediate intervention needed",
      "Annual follow-up recommended",
      "Consider specialist referral if symptoms persist",
    ];
  } else if (
    resultLower.includes("tumor") ||
    resultLower.includes("pituitary")
  ) {
    return [
      "Immediate medical consultation recommended",
      "Additional imaging studies may be required",
      "Multidisciplinary team review suggested",
      "Consider specialist referral",
    ];
  } else {
    return [
      "Clinical correlation recommended",
      "Follow up with healthcare provider",
      "Consider additional diagnostic tests",
      "Monitor symptoms if any",
    ];
  }
};

/**
 * Download a PDF report for a medical analysis
 * @param {Object} result - The analysis result data
 * @param {string} filename - Optional custom filename
 */
export const downloadMedicalReport = async (result, filename) => {
  try {
    const pdf = await generateMedicalPDF(result);
    const defaultFilename = `medical_report_${
      result.patientId || result.id
    }_${formatNepaliTime(new Date(), {
      year: "numeric",
      month: "2-digit",
      day: "2-digit",
      hour: undefined,
      minute: undefined,
      timeZoneName: undefined,
    }).replace(/\//g, "-")}.pdf`;

    pdf.save(filename || defaultFilename);
  } catch (error) {
    console.error("Error downloading PDF:", error);
    throw error;
  }
};

/**
 * Get analysis type display name
 * @param {string} modelType - The model type
 * @returns {string} Display name for the analysis type
 */
const getAnalysisTypeDisplay = (modelType) => {
  switch (modelType) {
    case "tumor":
      return "Brain Tumor Detection";
    case "chest_xray":
      return "Pneumonia Detection";
    default:
      return "Medical Analysis";
  }
};
